cooltools/expected.py

```diff
@@ -920,6 +920,170 @@ def blocksum_asymm(
+            )
+        if chrom is not None:
+            name = f"{chrom}:" + name
+        names.append(name)
+
+    return names, processed_regions
+
+
+def mat_expected(
+    heatmap,
+    regions=None,
+    raw_map=None,
+    *,
+    exclude_diag=2,
+    chrom=None,
+    resolution=1,
+    offset_bp=0,
+):
+    """
+    Calculates open2c-formatted expected from a heatmap.
+    Optionally adds chromosomes, and converts region names to meaningful names.
+
+    Parameters
+    ----------
+
+    heatmap: 2D array
+        Heatmap to calculate expected (expected goes to balanced.sum)
+    regions: [(st, end), ((st1, end1), (st2, end2))]
+        on-diagonal or off-diagonal regions to calculate expected
+    raw_map: 2D array or None (optional; default=None)
+        Raw heatmap to populate count.sum
+    exclude_diags: int (optional; default=2)
+        Number of diagonals to exclude. Resulting diags are not filled with NaNs.
+    chrom: str or None (optional; default=None)
+        If set to str, would append chrom to region name to obtain UCSC-style strings
+    resolution: int (optional; default=1)
+        If set to a number, would assume this is a size of a pixel.
+        Helpful for obtaining UCSC-style regions that are meaningful.
+    offset_bp: int (optional; default=0)
+        Add this offset to start/end for region name only.
+        Helpful if your heatmap does not start at the beginning of a chromosome
+
+    Caveats
+    -------
+
+    For symmetric regions, every diagonal is counted once.
+
+    For assymetric regions that do not cross main diagonal,
+    each diagonal is counted once as well.
+
+    For assymetric regions that cross main diagonal (RLY?)
+    the "overhang" would be counted twice. Also, why would you want that?
+    """
+
+    heatmap = np.array(heatmap, dtype=float)
+    M = len(heatmap)
+
+    if regions is None:
+        regions = [(0, M)]
+
+    names, regions = _preprocess_regions(
+        regions,
+        M=M,
+        chrom=chrom,
+        resolution=resolution,
+        offset_bp=offset_bp,
+    )
+
+    heatmap[~np.isfinite(heatmap)] = 0
+    mask = np.sum(heatmap, axis=0) == 0  # valid diag mask
+    exclude_diag = 2
+
+    valid = np.ones_like(heatmap, dtype=bool)
+    valid[mask] = False
+    valid[:, mask] = False
+
+    ar = np.arange(len(heatmap), dtype=np.int32)
+    diag_mask = np.abs(ar[:, None] - ar[None, :])
+    diag_mask[~valid] = -1
+
+    dfs = []
+    for name, reg in zip(names, regions):
+        (st1, ed1), (st2, ed2) = reg
+        subhm = heatmap[st1:ed1, st2:ed2].flatten()
+        subdiag = diag_mask[st1:ed1, st2:ed2].flatten()
+
+        mask = subdiag >= exclude_diag
+        subhm = subhm[mask]
+        subdiag = subdiag[mask]
+
+        n_valid = np.bincount(subdiag)
+        balanced_sum = np.bincount(subdiag, weights=subhm)
+        diag = np.arange(len(n_valid), dtype=int)
+
+        mask2 = n_valid > 0
```

Comment:
Are we dropping diagonals with 0 valid pixels? I thought the expected format should include a record for every relevant diagonal.

Reply:
Hm. I will think about it. Not 100% sure why I wrote this.

Reply:
just saw the comment - I've noticed that as well ... I think we should NaN-fill those